### PR TITLE
[Transaction] Fix transaction buffer client channel not active problem.

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
when channel is not active, it will not invalidate the `ClientCnx`. 
## implement
check channel active.
### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java [hunk 1]
```diff
@@ -116,24 +124,26 @@ public TransactionBufferHandlerImpl(PulsarClient pulsarClient,
...
-                        pendingRequests.remove(requestId);
+                        cb.completeExceptionally(
+                                new PulsarClientException.LookupException(topic + " endTxn channel is not active"));
                         op.recycle();
                     }
                 } else {
+                    log.error("endTxn error topic: [{}]", topic, throwable);
                     cache.invalidate(topic);
                     cb.completeExceptionally(throwable);
-                    pendingRequests.remove(requestId);
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java:136]: Why are you removing this line?

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java (hunk 1 above)
Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java:136]: if run here, the request don't put in `pendingRequests`. so remove this line

Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java:136]: makes sense, thanks

**Review by eolivelli — APPROVED** [review]
+1

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java [hunk 2]
```diff
@@ -116,24 +124,26 @@ public TransactionBufferHandlerImpl(PulsarClient pulsarClient,
...
         OpRequestSend op = OpRequestSend.create(requestId, topic, cmd, cb);
-        pendingRequests.put(requestId, op);
-        cmd.retain();
         try {
             cache.get(topic).whenComplete((clientCnx, throwable) -> {
                 if (throwable == null) {
-                    try {
+                    if (clientCnx.ctx().channel().isActive()) {
                         clientCnx.registerTransactionBufferHandler(TransactionBufferHandlerImpl.this);
+                        synchronized (TransactionBufferHandlerImpl.class) {
```
Comment (lhotari) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java:132]: It seems odd to use the Class object as an object monitor. Is this intentional? 
Did you mean to have `synchronized (TransactionBufferHandlerImpl.this) {` instead?

Comment (congbobo184) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/TransactionBufferHandlerImpl.java:132]: it my mistake, good catch!

**Pull request merged**